leshan-tl-cf-server-coap-oscore/src/main/java/org/eclipse/leshan/transport/californium/server/OscoreContextCleaner.java

```diff
@@ -58,7 +58,7 @@ public void updated(RegistrationUpdate update, Registration updatedReg, Registra
     @Override
     public void unregistered(Registration registration, Collection<Observation> observations, boolean expired,
             Registration newReg) {
-        if (registration.getClientTransportData().getIdentity() instanceof OscoreIdentity) {
+        if (registration.getClientTransportData().getIdentity() instanceof OscoreIdentity && newReg == null) {
             removeContext(((OscoreIdentity) registration.getClientTransportData().getIdentity()).getRecipientId());
```

Comment:
I see why it doesn't work without that.
but I'm not so sure about the solution.

I think there is case where we want to remove the context :thinking: 
Maybe we should not remove it only old registration and new registration use same `OscoreIdentity` ? 

(note that first argument of unregistered method is the old registration)

Reply:
Nice!

Not sure what your preferences are for nesting or verbosity, but what about something like:
```
    @Override
    public void unregistered(Registration registration, Collection<Observation> observations, boolean expired,
            Registration newReg) {
        if (registration.getClientTransportData().getIdentity() instanceof OscoreIdentity) {
            boolean matchingIdentity = registration.getClientTransportData().getIdentity()
                    .equals(newReg.getClientTransportData().getIdentity());
            if (!matchingIdentity) {
                removeContext(((OscoreIdentity) registration.getClientTransportData().getIdentity()).getRecipientId());
            }
        }
    }
```

Reply:
I think there is potential NPE issue here. 

Generally I privilege readability. (but this is very subjective :sweat_smile:)
Also, i feel exist fast help for reading.


So maybe something like that ?
``` java
LwM2mIdentity unregisteredIdentity = registration.getClientTransportData().getIdentity();

// Not an OSCORE identity : nothing to clear
if (!(unregisteredIdentity instanceof OscoreIdentity))
    return;

// This is a Re-Registration and client use same OSCORE Identity : nothing to clear
if (newReg != null && unregisteredIdentity.equals(newReg.getClientTransportData().getIdentity())) {
    return;
}

// else
removeContext(((OscoreIdentity) unregisteredIdentity).getRecipientId());
```

Reply:
> I think there is potential NPE issue here. (if negReg is null)

Yeah, I noticed and updated the comment, but probably it updated after you saw it:
```
boolean matchingIdentity = newReg != null && registration.getClientTransportData().getIdentity()
                .equals(newReg.getClientTransportData().getIdentity());
```

> Generally I privilege readability. (but this is very subjective 😅) Also, i feel exist fast help for reading.
> 
> So maybe something like that ?

Looks good! I have updated the PR aligned with that now.